src/GraphQL/Extensions/ObjectExtensions.cs

```diff
@@ -28,175 +28,268 @@ public static class ObjectExtensions
+                {
+                    if (fields[j].MemberName == fieldName)
+                        throw new InvalidOperationException($"Two fields within graph type '{graphType.Name}' were mapped to the same member '{fieldName}'.");
+                }
+                // add to list
+                fields[i] = (fieldType.Name, fieldName, resolvedType);
+            }
 
-            for (int i = 0; i < ctorParameters.Length; ++i)
+            // find best constructor to use, with preference to the constructor with the most parameters
+            var (bestConstructor, ctorParameters) = _types.GetOrAdd(
+                clrType,
+                static clrType =>
+                {
+#pragma warning disable IL2067 // Target parameter argument does not satisfy 'DynamicallyAccessedMembersAttribute' in call to target method. The parameter of method does not have matching annotations.
+                    var constructor = AutoRegisteringHelper.GetConstructor(clrType);
+#pragma warning restore IL2067 // Target parameter argument does not satisfy 'DynamicallyAccessedMembersAttribute' in call to target method. The parameter of method does not have matching annotations.
+                    var parameters = constructor.GetParameters();
+                    return (constructor, parameters);
+                });
+
+            // pull out parameters that are applicable for that constructor
+            var memberCount = fields.Length;
+            var ctorFields = ctorParameters.Length > 0
+                ? new ReflectionInfo.CtorParameterInfo[ctorParameters.Length]
+                : Array.Empty<ReflectionInfo.CtorParameterInfo>();
+            for (var i = 0; i < ctorParameters.Length; i++)
             {
-                var fieldName = usedKeys![i];
-                if (fieldName is not null)
+                var ctorParam = ctorParameters[i];
+                // look for a field that matches the constructor parameter name
+                var index = Array.FindIndex<(string, string? MemberName, IGraphType)>(fields, 0, fields.Length, x => string.Equals(x.MemberName, ctorParam.Name, StringComparison.OrdinalIgnoreCase));
+                if (index == -1)
                 {
-                    var fieldType = inputGraphType.Fields.Find(fieldName)?.ResolvedType
-                        ?? throw new InvalidOperationException($"Could not get ResolvedType for field '{fieldName}' of type '{inputGraphType}'.");
-                    object? arg = GetPropertyValue(values[i], ctorParameters[i].ParameterType, fieldType);
-                    ctorArguments[i] = arg;
+                    if (ctorParam.IsOptional)
+                        ctorFields[i] = new() { ParameterInfo = ctorParam };
+                    else
+                        throw new InvalidOperationException($"Cannot find field named '{ctorParam.Name}' on graph type '{graphType.Name}' to fulfill constructor parameter for type '{clrType.GetFriendlyName()}'.");
```

Comment:
I don't want to do it in this PR because for scoped schemas, it would seriously slow down initialization of the schema.  However with #3761 I'm adding a global switch to enable/disable precompilation and I'll configure it to compile `ToObject` upon initialization of the schema (assuming precompilation is enabled), revealing any issues upon initialization of the schema.

Reply:
Ugh it can't precompile during the `GraphType.Initialize` method because the `ResolvedType` property is not filled out for each graph type yet.  I can make it pull the reflection information ahead of time based on `TSourceType`, which would catch these errors, but then I have to expose the reflection information class as public and add additional public methods.  Maybe this should be done as a separate PR to both of these PRs.

Reply:
I talked in general, about new validations added with every major version. It's very frustrating to hunt runtime failures that can be caught only when executing specific queries. On every major upgrade, we run our ent-to-end tests and find that something that worked before started throwing in the new version. Then, we go over our schema (which is huge) and look for similar patterns to find all the places that weren't covered by tests.
It would be great to have a configuration flag to run all the validations on startup.

Reply:
Completely agree